# A systematic machine learning and data type comparison yields robust metagenomic predictors of infant age, sex, breastfeeding, antibiotic usage, country of origin, and delivery type

You can find here the associated code for our publication on machine learning in the microbiome. To see how we processed the metagenomic data, check out the "metagenomic_pipeline" folder. To see the machine learning, take a look at the "analysis_pipeline."

You can read our abstract below:

The microbiome is a new frontier for building predictors of human phenotypes. However, the practice of machine learning of human phenotype as a function of microbiome features is fraught with issues of reproducibility. A threat to reproducibility is the wide range of analytic models and microbiome data types available. We aimed to build robust metagenomic predictors of host phenotype by comparing prediction performances and biological interpretation across 8 machine learning methods and 5 different types of metagenomic data. Using 1,570 samples from 300 infants, we fit 7,865 models for 6 host phenotypes. We demonstrate the dependence of accuracy on algorithm choice and feature definition in microbiome data, and propose a framework for building microbiome-derived indicators of host phenotype. We additionally identified biological features predictive of age, sex, breastfeeding status, historical antibiotic usage, country of origin, and delivery type. Our complete results can be viewed at http://apps.chiragjpgroup.org/ubiome_predictions/.


